app/src/main/java/org/mozilla/rocket/download/DownloadInfoLiveData.kt

```diff
@@ -0,0 +1,289 @@
+import android.database.Cursor
+import android.os.Handler
+import android.support.v4.content.LocalBroadcastManager
+import org.mozilla.focus.download.DownloadInfo
+import org.mozilla.focus.download.DownloadInfoManager
+import org.mozilla.focus.utils.CursorUtils
+import org.mozilla.threadutils.ThreadUtils
+
+class DownloadInfoLiveData(private val context: Context) : LiveData<DownloadInfoBundle>() {
+
+    private val downloadInfoBundle: DownloadInfoBundle = DownloadInfoBundle(ArrayList(), -1, -1)
+
+    private var handler: Handler? = null
+    private var itemCount = 0
+    var isOpening = false
+    private var isLoading = false
+    private var isLastPage = false
+
+    private val broadcastReceiver = object : BroadcastReceiver() {
+
+        override fun onReceive(context: Context, intent: Intent) {
+            if (intent.action == DownloadManager.ACTION_DOWNLOAD_COMPLETE) {
+                val id = intent.getLongExtra(DownloadManager.EXTRA_DOWNLOAD_ID, 0L)
+                if (id > 0) {
+                    DownloadInfoManager.getInstance().queryByDownloadId(id, updateListener)
+                }
+            } else if (intent.action == DownloadInfoManager.ROW_UPDATED) {
+                val id = intent.getLongExtra(DownloadInfoManager.ROW_ID, 0L)
+                if (id > 0) {
+                    DownloadInfoManager.getInstance().queryByRowId(id, updateListener)
+                }
+            }
+        }
+    }
+
+    var updateListener = { downloadInfoList: List<DownloadInfo> ->
+        for (downloadInfo in downloadInfoList) {
+            updateItem(downloadInfo)
+        }
+    }
+
+    private val runningDownloadIds: LongArray
+        get() {
+            val ids = downloadInfoBundle.list
+                    .filter { it.status == DownloadManager.STATUS_RUNNING || it.status == DownloadManager.STATUS_PENDING }
+                    .map { it.downloadId }
+            val array = LongArray(ids.size)
+            for (i in array.indices) {
+                array[i] = ids[i]
+            }
+
+            return array
+        }
+
+    private val isDownloading: Boolean
+        get() {
+            return downloadInfoBundle.list.any { it.status == DownloadManager.STATUS_RUNNING || it.status == DownloadManager.STATUS_PENDING }
+        }
+
+    override fun onActive() {
+        super.onActive()
+        LocalBroadcastManager.getInstance(context).registerReceiver(broadcastReceiver, IntentFilter(DownloadInfoManager.ROW_UPDATED))
+        context.registerReceiver(broadcastReceiver, IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE))
+        val runningIds = runningDownloadIds
+        if (!runningIds.isEmpty()) {
+            updateRunningItems()
+        }
+    }
+
+    override fun onInactive() {
+        super.onInactive()
+        LocalBroadcastManager.getInstance(context).unregisterReceiver(broadcastReceiver)
+        context.unregisterReceiver(broadcastReceiver)
+        cleanUp()
+    }
+
+    fun loadMore(init: Boolean) {
```

Comment:
methods like this make LiveData having too many responsibilities, we should move this to a presenter. If you like MVVM, put them in VM. If you want MVVM, we need to put classes related to this feature in one package.